### PR TITLE
8027 QML & Web generated audio now follow the client audio device selection

### PR DESCRIPTION
**Pull request by luiscuenca**
This PR makes QML & Web surfaces responsive to changes on the client audio selection. 

https://highfidelity.fogbugz.com/f/cases/8027

...

## TEST PLAN
== WEB ENTITIES
- Go to the Sandbox or other domain where you can create entities.
- Create a new web entity.
- Change its url to youtube.com.
- Select any video and start watching ( they should reproduce sound using the current device ).
- Go to the audio menu and change the output audio device.
- The video will reproduce sound using the selected device.
- Delete the web entity.

== WEB BROWSER
- Open the interface web browser (CTRL + SHIFT + B).
- Change its url to youtube.com.
- Select any video and start watching ( they should reproduce sound using the current device ).
- Go to the audio menu and change the output audio device.
- The video will reproduce sound using the selected device.
- Close the browser.

== QML MULTIMEDIA
- Open the script window (Edit -> Running Scripts... or CTRL + J) 
- Click on "FROM URL" and add the link https://s3-us-west-1.amazonaws.com/hifio/qmlmediatest.js
- An audio loaded from a QML file should start playing using the selected output device. 
- Go to the audio menu and change the output audio device.
- The audio will reproduce sound using the selected device.

...

## KNOWN ISSUES
- Only media DOM elements will be updated. Embedded media elements won’t (youtube videos on hifidelity.io).
- Every time a HTML media elements gets added to the DOM, the default audio output device is automatically asign as their output audio. This means that every time the page loads or changes its content, all the DOM elements will need to be updated with the selected audio device. In some cases this changes are difficult or impossible to cast.( a <video> is added 3 seconds after page loads ) 
- QML video only accepts the wmv file format and doesn't render propertly (only audio).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4166/HighFidelity-Beta-PR11600-df09ec6891b04715b94324a5da4afce9e75177a8.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4169/HighFidelity-Beta-PR11600-23c689be1bbaf0a3bd130602c080c5bc1aba2c47.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4169/HighFidelity-Beta-PR11600-23c689be1bbaf0a3bd130602c080c5bc1aba2c47.exe

File: interface/resources/html/createGlobalEventBridge.js [hunk 1]
```diff
@@ -33,6 +33,57 @@ var EventBridge;
         // replace the TempEventBridge with the real one.
         var tempEventBridge = EventBridge;
         EventBridge = channel.objects.eventBridge;
+		EventBridge.onChangeOutputAudioDeviceTo.connect(function(deviceName) {
```
Comment (jherico) [inline interface/resources/html/createGlobalEventBridge.js]: use spaces, not tabs.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -4268,6 +4268,46 @@ void Application::init() {
...
+    });
+
+    EntityTreePointer tree = getEntities()->getTree();
+
+    // Make every web entity aware of output audio device changes (Connecting at creation time) 
+    connect(tree.get(), &EntityTree::addingEntity, this, [this](const EntityItemID& entityID) {
+        EntityItemPointer newEntity = getEntities()->getTree()->findEntityByEntityItemID(entityID);
+        auto etype = newEntity->getType();
+        if (etype == EntityTypes::Web) {            
+            connect(this, &Application::changeOutputAudioDevice, this, [newEntity](QString deviceName) {
```
Comment (jherico) [inline interface/src/Application.cpp]: You've captured a shared pointer in a lambda used in a connection.  I'm pretty sure this means that the shared pointer here can never leave scope and the entity will never be destroyed.  You can avoid this by converting newEntity to a weak pointer and then capturing that.  in the lambda you will need to lock the weak pointer to re-aqcuire a shared pointer and test if it's valid before calling `changeOutputAudioDeviceTo`

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -4268,6 +4268,46 @@ void Application::init() {
...
+                newEntity->changeOuputAudioDeviceTo(deviceName);
+            });
+            connect(newEntity.get(), &EntityItem::onUpdateOutputAudioDevice, this, [newEntity]() {
```
Comment (jherico) [inline interface/src/Application.cpp]: same.... don't capture shared pointers in lambdas.

File: interface/src/Application.h [hunk 4]
```diff
@@ -312,6 +312,8 @@ class Application : public QApplication,
     void uploadRequest(QString path);
     void receivedHifiSchemeURL(const QString& url);
 
+    void changeOutputAudioDevice(QString deviceName);
```
Comment (jherico) [inline interface/src/Application.h]: prefer `const T&` for function parameters.

File: libraries/entities-renderer/src/RenderableWebEntityItem.cpp [hunk 5]
```diff
@@ -265,7 +265,12 @@ bool WebEntityRenderer::buildWebSurface(const TypedEntityPointer& entity) {
             QCoreApplication::postEvent(_webSurface->getWindow(), touchEvent);
         }
     });
-
+    
+    // //Connect Audio Events
+    connect(this->getEntity().get(), &EntityItem::onChangeOutputAudioDeviceTo, this, [this](QString deviceName) {
```
Comment (jherico) [inline libraries/entities-renderer/src/RenderableWebEntityItem.cpp]: `entity` is in scope here, so try to use that.  We want to minimize the number of places we reference the entity as a member as opposed to a parameter.

Also, `const QString& deviceName`

File: libraries/entities-renderer/src/RenderableWebEntityItem.cpp [hunk 6]
```diff
@@ -265,7 +265,12 @@ bool WebEntityRenderer::buildWebSurface(const TypedEntityPointer& entity) {
...
+        _webSurface->changeOutputAudioDeviceTo(deviceName);
+    });
+    connect(_webSurface.data(), &OffscreenQmlSurface::onOutputAudioDeviceNeedsUpdate, this->getEntity().get(), &EntityItem::updateOuputAudioDevice);
```
Comment (jherico) [inline libraries/entities-renderer/src/RenderableWebEntityItem.cpp]: same

File: libraries/entities/src/EntityItem.cpp [hunk 7]
```diff
@@ -3006,3 +3006,11 @@ void EntityItem::somethingChangedNotification() {
         }
     });
 }
+
+void EntityItem::updateOuputAudioDevice() {
```
Comment (jherico) [inline libraries/entities/src/EntityItem.cpp]: We actually want to avoid putting details about 'rendering' (including audio rendering) inside entities.  It would be preferable if we could move these signals and functions to the base class for entity rendering, `EntityRenderer`

File: libraries/ui/src/ui/OffscreenQmlSurface.cpp [hunk 8]
```diff
@@ -248,6 +248,22 @@ size_t OffscreenQmlSurface::getUsedTextureMemory() {
     return offscreenTextures.getUsedTextureMemory();
 }
 
+void OffscreenQmlSurface::changeOutputAudioDeviceTo(QString deviceName) {
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.cpp]: `const QString&`

File: libraries/ui/src/ui/OffscreenQmlSurface.h [hunk 9]
```diff
@@ -98,14 +98,18 @@ class OffscreenQmlSurface : public QObject {
 public slots:
     void onAboutToQuit();
     void focusDestroyed(QObject *obj);
+    void changeOutputAudioDeviceTo(QString deviceName);
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.h]: `const QString&`

File: libraries/ui/src/ui/OffscreenQmlSurface.h [hunk 10]
```diff
@@ -98,14 +98,18 @@ class OffscreenQmlSurface : public QObject {
...
     // event bridge
 public slots:
     void emitScriptEvent(const QVariant& scriptMessage);
     void emitWebEvent(const QVariant& webMessage);
+    void outputAudioDeviceNeedsToUpdate();
 signals:
     void scriptEventReceived(const QVariant& message);
     void webEventReceived(const QVariant& message);
+    void onChangeOutputAudioDeviceTo(QString deviceName);
+    void onOutputAudioDeviceNeedsUpdate();
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.h]: This naming is counter-intuitive.  Signals typically do not have names like with `onXXX`.  Instead that usually indicates a slot handling signal `XXX`.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4240/HighFidelity-Beta-PR11600-ecdf77a1fdf0f4e25d995f320287ac49271674c2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4240/HighFidelity-Beta-PR11600-ecdf77a1fdf0f4e25d995f320287ac49271674c2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4244/HighFidelity-Beta-PR11600-4a1fa76c96f68de50c3c62a1c7e9f545f4f78afe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4244/HighFidelity-Beta-PR11600-4a1fa76c96f68de50c3c62a1c7e9f545f4f78afe.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4259/HighFidelity-Beta-PR11600-ae2c713d5a81456fc75b030550ed1bf2602c0a1d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4259/HighFidelity-Beta-PR11600-ae2c713d5a81456fc75b030550ed1bf2602c0a1d.exe

File: libraries/ui/src/ui/OffscreenQmlSurface.cpp [hunk 11]
```diff
@@ -296,6 +296,22 @@ size_t OffscreenQmlSurface::getUsedTextureMemory() {
...
 
+void OffscreenQmlSurface::changeOutputAudioDeviceTo(const QString& deviceName) {
+    if (QThread::currentThread() != thread()) {
+        QMetaObject::invokeMethod(this, "changeOutputAudioDeviceTo", Qt::QueuedConnection, Q_ARG(const QString&, deviceName));
+    }  else {
+        emit setOutputAudioDeviceTo(deviceName);
+    }
+}
+
+void OffscreenQmlSurface::onOutputAudioDeviceNeedsToUpdate() {
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.cpp]: this naming is still confusing.  Please calll this something like `forceAudioOutputDeviceUpdate`

File: libraries/ui/src/ui/OffscreenQmlSurface.cpp [hunk 12]
```diff
@@ -296,6 +296,22 @@ size_t OffscreenQmlSurface::getUsedTextureMemory() {
...
+    if (QThread::currentThread() != thread()) {
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.cpp]: you don't need to do this thread test and self-invocation.  Instead just invoke the signal directly:  `QMetaObject::invokeMethod(this, "outputAudioDeviceNeedsUpdate");` ... allowing the auto-conenction type will do a direct signal emission if on the same thread and a queued signal emission if called from another thread.

File: interface/src/Application.cpp [hunk 13]
```diff
@@ -4263,6 +4263,38 @@ void Application::init() {
             avatar->setCollisionSound(sound);
         }
     }, Qt::QueuedConnection);
+
+    connectOutputAudioEmitter();
+}
+
+void Application::connectOutputAudioEmitter() {
```
Comment (jherico) [inline interface/src/Application.cpp]: This logic only seems to handle one offscreen surface... the primary UI.  This means it won't work for the tablet, or for any other QML surfaces like Web Entities or Web Overlays.  

Would it not make more sense for the connection functionality to happen inside OffscreenQmlSurface, so that it can connect directly to the AudioClient signals and take action as necessary?

File: interface/resources/html/createGlobalEventBridge.js [hunk 14]
```diff
@@ -33,6 +33,57 @@ var EventBridge;
         // replace the TempEventBridge with the real one.
         var tempEventBridge = EventBridge;
         EventBridge = channel.objects.eventBridge;
+        EventBridge.setOutputAudioDeviceTo.connect(function(deviceName) {
```
Comment (jherico) [inline interface/resources/html/createGlobalEventBridge.js]: This code updates sound output for web content, but not for QML content that emits audio.  Is the intent to do this in another PR, or is there otherwise a plan for this?

File: libraries/entities-renderer/src/RenderableWebEntityItem.cpp [hunk 15]
```diff
@@ -265,6 +265,10 @@ bool WebEntityRenderer::buildWebSurface(const TypedEntityPointer& entity) {
             QCoreApplication::postEvent(_webSurface->getWindow(), touchEvent);
         }
     });
+    
+    // Connect Audio Events
+    QObject::connect(renderer.data(), &EntityTreeRenderer::setOutputAudioDevice, _webSurface.data(), &OffscreenQmlSurface::changeOutputAudioDeviceTo);
```
Comment (jherico) [inline libraries/entities-renderer/src/RenderableWebEntityItem.cpp]: these connection must have corresponding disconnection logic in the `destroyWebSurface` method.  See the disconnection code for the other entity tree renderer signals.

File: libraries/ui/src/ui/OffscreenQmlSurface.h [hunk 16]
```diff
@@ -102,14 +102,18 @@ class OffscreenQmlSurface : public QObject {
...
 
     // event bridge
 public slots:
     void emitScriptEvent(const QVariant& scriptMessage);
     void emitWebEvent(const QVariant& webMessage);
+    void onOutputAudioDeviceNeedsToUpdate();
 signals:
     void scriptEventReceived(const QVariant& message);
     void webEventReceived(const QVariant& message);
+    void setOutputAudioDeviceTo(const QString& deviceName);
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.h]: please rename to `void audioOutputDeviceChanged(const QString& deviceName)`

File: libraries/ui/src/ui/OffscreenQmlSurface.h [hunk 17]
```diff
@@ -102,14 +102,18 @@ class OffscreenQmlSurface : public QObject {
...
+    void outputAudioDeviceNeedsUpdate();
```
Comment (jherico) [inline libraries/ui/src/ui/OffscreenQmlSurface.h]: eliminate signal, integrate indrection logic from application.cpp directly into the renamed `onOutputAudioDeviceNeedsToUpdate`

File: interface/src/Application.cpp [hunk 18]
```diff
@@ -4263,6 +4263,38 @@ void Application::init() {
...
+    // Listen to audio device changes
+    auto audioIO = DependencyManager::get<AudioClient>();
+    connect(audioIO.data(), &AudioClient::deviceChanged, this, &Application::outputAudioDeviceChanged);
+       
+    // Connect the audio device selection to OffscreenUi (Browser)
+    auto offscreenUI = DependencyManager::get<OffscreenUi>();
+    connect(this, &Application::changeOutputAudioDevice, offscreenUI.data(), &OffscreenUi::changeOutputAudioDeviceTo);
+        
+    connect(offscreenUI.data(), &OffscreenUi::outputAudioDeviceNeedsUpdate, this, [this]() {
```
Comment (jherico) [inline interface/src/Application.cpp]: This indirection is unecessary.  You can put this logic into `onOutputAudioDeviceNeedsToUpdate` eliminating the need for a `outputAudioDeviceNeedsUpdate` signal.

File: interface/resources/html/createGlobalEventBridge.js (hunk 14 above)
Comment (luiscuenca) [reply interface/resources/html/createGlobalEventBridge.js]: I will incorporate QML audio on this PR

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4361/HighFidelity-Beta-PR11600-d965e952f0ed4061e3e0678541d1f77be224bb38.dmg

**Comment by jherico** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4364/HighFidelity-Beta-PR11600-d965e952f0ed4061e3e0678541d1f77be224bb38.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4373/HighFidelity-Beta-PR11600-e2a8b567c49ecaea7801d90348be999ed73317c4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4404/HighFidelity-Beta-PR11600-dad32dc74a5c9442f32c533b066579e958f65573.exe

**Comment by luiscuenca** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4409/HighFidelity-Beta-PR11600-dad32dc74a5c9442f32c533b066579e958f65573.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4409/HighFidelity-Beta-PR11600-dad32dc74a5c9442f32c533b066579e958f65573.exe

**Comment by Nex-Pro** [comment]
Testing win10

**Comment by Nex-Pro** [comment]
### PASS on all three tests.

Changing audio devices is instantaneous and reliable.

File: libraries/ui/src/QmlWindowClass.cpp [hunk 19]
```diff
@@ -128,7 +128,7 @@ void QmlWindowClass::initQml(QVariantMap properties) {
             object->setProperty(SOURCE_PROPERTY, _source);
 
             const QMetaObject *metaObject = _qmlWindow->metaObject();
-            // Forward messages received from QML on to the script
+            // //Forward messages received from QML on to the script
```
Comment (ZappoMan) [inline libraries/ui/src/QmlWindowClass.cpp]: this extra comment is weird

Comment (luiscuenca) [reply libraries/ui/src/QmlWindowClass.cpp]: Yes it is, but it has a "logical" explanation. I was modifying the QML file on notepad++. That extra comment tells VS that the project was modified and it needs to update the QML. 
I am starting to modify the QML files on VS (no extra tabs).

**Comment by luiscuenca** [comment]
Closing if favor of https://github.com/highfidelity/hifi/pull/11706
